### ISSUE TITLE
[HOLD on #378314][$500] [Held requests] Request on hold becomes unheld when approving only unheld request

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#39338]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.58-4
**Reproducible in staging?:** Y
**Reproducible in production?:** N
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N/A
**Issue reported by:** Applause - Internal Team

**Issue found when executing PR** https://github.com/Expensify/App/pull/33124
## Action Performed:
Precondition:
- Admin and employee are in the Collect workspace
1. [Employee] Request money from the workspace chat
2. [Employee] Create two money requests
3. [Employee] Hold one of the requests and leave the other one unheld
4. [Admin] Go to expense report
5. [Admin] Click Approve button
5. [Admin] Approve the unheld request

## Expected Result:
The request on hold will be moved to a new report

## Actual Result:
Request on hold becomes unheld when approving only unheld request

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/c6d765a1-acc8-456e-ba42-bc556846f263

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01d6a3e193d64c10a4</li>
        <li>Upwork Job ID: 1775018433509507072</li>
        <li>Last Price Increase: 2024-04-09</li>
    </ul>
</details>

**Comment by lanitochka17** [comment]
We think that this bug might be related to #wave-collect - Release 1

**Comment by lanitochka17** [comment]
@NikkiWines  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by NikkiWines** [comment]
This definitely looks like a regression / unhandled case from https://github.com/Expensify/App/pull/33124. I've been trying to reproduce locally but the behavior is pretty buggy on `main` right now (crashing when holding an expense, approving the full total and not the held amount) so it's hard to see what's causing this particular staging behavior 

Pinged @robertjchen since he's got context on the original PR and might be able to lend some insight here

**Comment by NikkiWines** [comment]
Looks like - at least on dev - it's the `optimisticIOUReportData` from [here](https://github.com/Expensify/App/blob/2175753817eaaf40422078ac74753f4754415eec/src/libs/actions/IOU.ts#L4629) that is overwriting the `amount` with the full (held + non-held) amount, even though the `amount` is correctly being set in `optimisticApprovedReportAction` [here](https://github.com/Expensify/App/blob/2175753817eaaf40422078ac74753f4754415eec/src/libs/actions/IOU.ts#L4643-L4656).

**Comment by robertjchen** [comment]
Thanks for the report! Definitely related to the changes there: https://github.com/Expensify/App/pull/33124 and we'll be addressing it. Not a deploy blocker per the [sentiments](https://expensify.slack.com/archives/C036QM0SLJK/p1711507708986889?thread_ts=1711403751.963269&cid=C036QM0SLJK) here

**Comment by NikkiWines** [comment]
Sounds good, marking this as an external bug then so it can get worked on 👍

**Comment by nkdengineer** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Request on hold becomes unheld when approving only unheld request

### What is the root cause of that problem?
When we approve only unheld request, we don't create new expense report and report preview for the hold request. And BE also doesn't implement this like the expected here

> The request on hold will be moved to a new report


https://github.com/Expensify/App/blob/a4d01a87b242eef06af37db965bd135745e342e5/src/libs/actions/IOU.ts#L4623
### What changes do you think we should make in order to solve the problem?
Add a new variable like `shouldCreateNewPreview` and update it to `true` in this case.

https://github.com/Expensify/App/blob/a4d01a87b242eef06af37db965bd135745e342e5/src/libs/actions/IOU.ts#L4626-L4627

If it's true, we will move all hold request to new report. The main steps:
1. Create an optimistic expense report
2. Move all IOU actions hold transactions from old expense report to new expense report
3. Update the total of new expense report as the `expenseReport.total - expenseReport.unheldTotal`
4. Update the total of the old expense report as `expenseReport.unheldTotal`
5. Create an optimistic report preview for new expense report, the child request count is the number of hold requests, other fields can be updated accordingly base on hold requests
6. Update the old report preview base on upheld requests

And we also need to implement this in back-end as well.
### What alternative solutions did you explore? (Optional)
NA

**Comment by NikkiWines** [comment]
@alitoshmatov please review the proposals when you have a moment 🙇

**Comment by alitoshmatov** [comment]
Reviewing

**Comment by alitoshmatov** [comment]
So as far as I see it is an issue with backend, because when doing the same offline it is working as expected. Even if there is an issue in frontend we should first fix backend and proceed from that

https://github.com/Expensify/App/assets/59907218/ebaa89ea-88ae-43c1-a9a3-c473d899d512

**Comment by trjExpensify** [comment]
@robertjchen are you aware of this backend issue with the hold project?

**Comment by trjExpensify** [comment]
Hm, interesting! But @marcochavezf only has the currency issue remaining, no? I guess I don't know where this is being fixed, or if his earlier PRs shoul dhave.

**Comment by trjExpensify** [comment]
Same, Melv!

**Comment by trjExpensify** [comment]
@marcochavezf can you confirm if this issue is resolved with the PRs that have come for [#378314](https://github.com/Expensify/Expensify/issues/378314)?

**Comment by marcochavezf** [comment]
Yup, this issue was fixed in this [PR](https://github.com/Expensify/App/pull/40065)

**Comment by trjExpensify** [comment]
Cool, once deployed we close this out then. :)

**Comment by NikkiWines** [comment]
Not overdue; waiting on https://github.com/Expensify/App/pull/40065, which is currently on staging.

**Comment by trjExpensify** [comment]
Hit prod two days ago, closing!

**Comment by nkdengineer** [comment]
@NikkiWines @trjExpensify We have this bug https://github.com/Expensify/App/issues/41652 here this is the part of this issue since we don't implement the optimistic data when we approve partial. Can you please take a look and re-open this issue.

**Comment by NikkiWines** [comment]
> @NikkiWines @trjExpensify We have this bug https://github.com/Expensify/App/issues/41652 here this is the part of this issue since we don't implement the optimistic data when we approve partial. Can you please take a look and re-open this issue.

Been focused on some other urgent issues but I will look into this during the week. From a quick glance it looks like there's some activity in the linked issue for resolving the issue as well. Re-opening for now so it doesn't fall off my radar.

**Comment by trjExpensify** [comment]
I think we can use the other issue for this follow-up. @nkdengineer I've assigned you for the C+ review on that.

**Comment by NikkiWines** [comment]
Sounds good, re-closing this then
